airbyte-workers/src/main/java/io/airbyte/workers/temporal/PartialSuccessTemporalAttemptExecution.java

```diff
@@ -0,0 +1,221 @@
+ * SOFTWARE.
+ */
+
+package io.airbyte.workers.temporal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.functional.CheckedConsumer;
+import io.airbyte.commons.functional.CheckedSupplier;
+import io.airbyte.scheduler.models.JobRunConfig;
+import io.airbyte.workers.Worker;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/*
+ * This class allow a worker to run multiple times. In addition to the functionality in
+ * TemporalAttemptExecution it takes a predicate to determine if the output of a worker constitutes
+ * a complete success or a partial one. It also takes a function that takes in the input of the
+ * previous run of the worker and the output of the last worker in order to generate a new input for
+ * that worker.
+ */
+public class PartialSuccessTemporalAttemptExecution<INPUT, OUTPUT> implements Supplier<List<OUTPUT>> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PartialSuccessTemporalAttemptExecution.class);
+
+  private final CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplier;
+  private final Supplier<INPUT> inputSupplier;
+  private final CancellationHandler cancellationHandler;
+
+  private final Predicate<OUTPUT> shouldAttemptAgainPredicate;
+  private final BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction;
+  private final int maxRetriesCount;
+  private final TemporalAttemptExecutionFactory<INPUT, OUTPUT> temporalAttemptExecutionFactory;
+  private final Path workspaceRoot;
+  private final JobRunConfig jobRunConfig;
+
+  public PartialSuccessTemporalAttemptExecution(Path workspaceRoot,
+                                                JobRunConfig jobRunConfig,
+                                                CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplier,
+                                                Supplier<INPUT> initialInputSupplier,
+                                                CancellationHandler cancellationHandler,
+                                                Predicate<OUTPUT> shouldAttemptAgainPredicate,
+                                                BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction,
+                                                int maxRetriesCount) {
+    this(
+        workspaceRoot,
+        jobRunConfig,
+        workerSupplier,
+        initialInputSupplier,
+        cancellationHandler,
+        shouldAttemptAgainPredicate,
+        computeNextAttemptInputFunction,
+        maxRetriesCount,
+        TemporalAttemptExecution::new);
+  }
+
+  @VisibleForTesting
+  PartialSuccessTemporalAttemptExecution(Path workspaceRoot,
```

Comment:
I recently learned that `protected` methods can be tested. See: https://junit.org/junit4/faq.html#atests_10. This means the annotation is unnecessary and access can be changed to `protected`

Reply:
I don't think this ctor should live here. if you want something like that for testing, you can put it in tests but not in src.

Reply:
agreed

Reply:
nvm - I forgot that `protected` still makes it package visible. Gonna go on a juice cleanse to get the Python out of my system.